runtime/src/main/scala/com/soundcloud/twinagle/ClientEndpointBuilder.scala

```diff
@@ -28,11 +40,36 @@ class ClientEndpointBuilder(
   ](
       endpointMetadata: EndpointMetadata
   ): Service[Req, Resp] = {
+    val retryFilter: Filter[Req, Resp, Req, Resp] = buildRetryFilter(endpointMetadata.isIdempotent)
     extension(endpointMetadata).toFilter andThen
+      retryFilter andThen
       new TracingFilter[Req, Resp](endpointMetadata) andThen
       new ProtobufClientFilter[Req, Resp](endpointMetadata.path) andThen
       new TwirpHttpClient andThen
       httpClient
 
   }
+
+  private def buildRetryFilter[Req, Resp](isIdempotent: Boolean) = { // TODO: Are these values reasonable? How configurable should they be?
+    val retryPolicy = if (isIdempotent) shouldRetry[Req, Resp] else defaultRetry
+    new RetryFilter[Req, Resp](
+      RetryPolicy.tries(3, retryPolicy),
+      HighResTimer.Default,
+      new NullStatsReceiver,
+      RetryBudget(60.seconds, 10, 0.2f)
+    )
+  }
+
+  // Retries all NonFatal exceptions
+  // TODO: Make a more informed decision about retrying
```

Comment:
Should the server implementor just provide a matcher?

Reply:
I think this could be a good idea for the clients to implement their own retry matcher. We could pass it in to the `HaberdasherClientProtobuf` or `HaberdasherClientJson` classes along with the extension, but make it an Option default None. What do you think?

Reply:
About this, should the retry matcher apply to all methods, or should each method accept a different matcher?

Reply:
I think I'd make it per-method, and default to `NonFatal` (so, this would rarely be specified)

Reply:
One issue about making it per method is that we would need to change the interface for the Service trait in order to allow an extra parameter for the retryMatcher in every method. Is that what we would want to do?

Reply:
Actually, if we did that, we could technically even provide retryMatchers on the server side as well. That way if a server hits a particular type of database exception that matches, it could automatically retry, which would be pretty cool